File: python/problems/leetcode/0121_best_time_to_buy_and_sell_stock/solution.py
"""LeetCode 121: Best Time to Buy and Sell Stock."""

from __future__ import annotations


class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        lo = 10**9
        best = 0
        for p in prices:
            lo = min(lo, p)
            best = max(best, p - lo)
        return best


if __name__ == "__main__":
    assert Solution().maxProfit([7, 1, 5, 3, 6, 4]) == 5
    assert Solution().maxProfit([7, 6, 4, 3, 1]) == 0
    print("0121 OK")

File: cpp/problems/leetcode/0121_best_time_to_buy_and_sell_stock/solution.cpp
// LeetCode 121
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int lo = INT_MAX, best = 0;
        for (int p : prices) {
            lo = min(lo, p);
            best = max(best, p - lo);
        }
        return best;
    }
};

int main() {
    vector<int> a{7, 1, 5, 3, 6, 4};
    assert(Solution().maxProfit(a) == 5);
    vector<int> b{7, 6, 4, 3, 1};
    assert(Solution().maxProfit(b) == 0);
    cout << "0121 OK" << endl;
    return 0;
}
